### PR TITLE
GH-15171: [C++] Pass std::string_view by value

### LINKED ISSUE #15171: [C++] Pass std::string_view by value, not by const-ref
### Describe the enhancement requested

See https://quuxplusone.github.io/blog/2021/11/09/pass-string-view-by-value/

### Component(s)

C++

### PR DESCRIPTION
**Pull request by ucasfl**
# Which issue does this PR close?

Closes #15171. 

# Rationale for this change

# What changes are included in this PR?

Pass std::string_view by value, not by const-ref
# Are these changes tested?

# Are there any user-facing changes?
No.

* Closes: #15171

### CODE REVIEW

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -49,6 +49,8 @@ endif()
 
 set(ARROW_VERSION "11.0.0-SNAPSHOT")
 
+set(CMAKE_EXPORT_COMPILE_COMMANDS 1) # Write compile_commands.json
```
Comment (pitrou) [inline cpp/CMakeLists.txt:52]: I don't think we want to enable this by default. You should pass this yourself to CMake using `-DCMAKE_EXPORT_COMPILE_COMMANDS=1`.

File: cpp/src/arrow/vendored/datetime/tz.cpp [hunk 2]
```diff
@@ -3597,19 +3597,17 @@ tzdb::locate_zone(const std::string& tz_name) const
 {
```
Comment (pitrou) [inline cpp/src/arrow/vendored/datetime/tz.cpp:3597]: This is a vendored file (as the file path suggests), so we shouldn't do such changes to it. Can you revert them?

**Review by pitrou — CHANGES_REQUESTED** [review]

File: cpp/CMakeLists.txt (hunk 1 above)
Comment (ucasfl) [reply cpp/CMakeLists.txt:52]: OK

File: cpp/src/arrow/vendored/datetime/tz.cpp (hunk 2 above)
Comment (ucasfl) [reply cpp/src/arrow/vendored/datetime/tz.cpp]: done.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
@ursabot please benchmark lang=C++

**Comment by cyb70289** [comment]
Rebased to fix lint error.

**Comment by cyb70289** [comment]
Looks Glib & Ruby CI error is not related? cc @kou  
https://github.com/apache/arrow/actions/runs/3935751955/jobs/6731706736#step:10:204

**Comment by kou** [comment]
Not related. It's sometimes happen. I couldn't reproduce it on local environment...

**Review by cyb70289 — APPROVED** [review]

**Comment by westonpace** [comment]
@ucasfl thanks for cleaning this up!

**Comment by jgehrcke** [comment]
Some curiosity-driven feedback CC @ucasfl @westonpace @pitrou. 

I read https://quuxplusone.github.io/blog/2021/11/09/pass-string-view-by-value/ with curiosity and was wondering if this patch here resulted in a measurable performance difference somewhere on the user-facing surface of Arrow.

Before Christmas I build a benchmark that I called `dataset-serialize`. Among others, it measures the time for dataset CSV serialization, writing to memory (/dev/shm). It also runs on bare metal. Both in combination yields quite some stability.

When this commit landed, said benchmark got faster by (roughly) three percent:

image

The plot/data can be further explored here: https://conbench.ursa.dev/benchmarks/bc6d37ffe2614078814efb01402f64dd/

Of course this is just correlation (once this was merged, the benchmark got a tad faster) and not necessarily causation (for corroborating this I understand too little of the impact of this change). But yeah, it's probably fair to say that it's not unlikely that this change here was responsible for the perf change.

**Pull request merged**